core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -131,15 +128,16 @@ private static boolean isMasterChannelException(TransportException exp) {
     /**
      * Send a shard failed request to the master node to update the
      * cluster state.
-     *  @param shardRouting       the shard to fail
-     * @param sourceShardRouting the source shard requesting the failure (must be the shard itself, or the primary shard)
+     * @param shardRouting       the shard to fail
+     * @param primaryTerm        the primary term associated with the primary shard when the shard is failed by the primary. Use 0L if the
```

Comment:
Using 0 as a non-value (because we serialize using out.writeVLong ) seems brittle to me. I rather use -1 and serialize using normal `out.writeLong`. With 0 we will always be asking ourselves if it's a valid value.  We can also potentially have two methods here - failReplica and failShard, where failReplica requires a valid primary term and failShard doesn't take one (but uses -1 internally) so users won't need to know about this implementation detail.